### PR TITLE
Add ResultStore Reporter to Crier

### PR DESCRIPTION
**Pull request by airbornepony**
Add a ResultStore Crier reporter

This reporter uploads results to the ResultStore API.

### CODE REVIEW

File: prow/cmd/crier/main.go
Comment (listx) [inline prow/cmd/crier/main.go:129]: s/resultstore report workers/ResultStore reporters/

File: prow/crier/reporters/resultstore/reporter.go
Comment (listx) [inline prow/crier/reporters/resultstore/reporter.go:57]: s/heurisitic/heuristic

File: prow/crier/reporters/resultstore/reporter.go
Comment (listx) [inline prow/crier/reporters/resultstore/reporter.go:80]: I'm surprised we don't have a similar function already. My guess is because we have https://github.com/kubernetes/test-infra/blob/3a69ac1572cf4a7878ced659de235310bee28d2b/prow/io/opener.go#L100 (Opener) for dealing with paths (not sure if any of that is relevant, because IIRC this reporter will not upload files to GCS itself at all but merely check that some files there exist and if so, report them to ResultStore).

File: prow/crier/reporters/resultstore/reporter_test.go
Comment (listx) [inline prow/crier/reporters/resultstore/reporter_test.go:218]: Nit: Maybe put this stuff up top to make it easier to discover?

File: prow/cmd/crier/main.go
Comment (airbornepony) [reply prow/cmd/crier/main.go:129]: The "report workers" is verbatim of the other flags. Shall I change them all to "reporters" instead?

File: prow/crier/reporters/resultstore/reporter.go
Comment (airbornepony) [reply prow/crier/reporters/resultstore/reporter.go]: Done.

File: prow/crier/reporters/resultstore/reporter.go
Comment (airbornepony) [reply prow/crier/reporters/resultstore/reporter.go:80]: Sorry, not sure if there's something actionable here?

File: prow/cmd/crier/main.go
Comment (airbornepony) [reply prow/cmd/crier/main.go]: Leaving "report workers" but renaming ResultStore with caps.

File: prow/crier/reporters/resultstore/reporter_test.go
Comment (airbornepony) [reply prow/crier/reporters/resultstore/reporter_test.go]: Done.

File: prow/crier/reporters/resultstore/reporter.go
Comment (listx) [reply prow/crier/reporters/resultstore/reporter.go:80]: Nope, just thinking out loud from my end.

Well, also the `isGCSBucket` function probably belongs outside of reporter.go in some other GCS-related area, but that is a very minor concern.

Comment (airbornepony) [reply prow/crier/reporters/resultstore/reporter.go:81]: Added a TODO. Moving this to util.go would allow a simpler impl. I did notice some other bits related to bucket identification (e.g. in config.go) but that doesn't seem the right place.

File: prow/resultstore/payload.go
Comment (ashleyrichter) [inline prow/resultstore/payload.go:49]: You don't need to set `id` or `name` fields on the resource protos when using batch upload API, you only need to set UploadRequest.id.

This is based on AIP spec for create request messages, but modified to suit the needs of our system.

https://google.aip.dev/133#user-specified-ids

"The {resource}_id field must exist on the request message, not the resource itself."
"The name field on the resource must be ignored."

This goes for all the resource protos, so you can delete a bunch of your code here.

(We could probably clarify the documentation around this, but it's hard to keep clarity across all these protos.)

File: prow/resultstore/payload.go
Comment (ashleyrichter) [inline prow/resultstore/payload.go:185]: As mentioned above, you can delete this line, but FYI, the proper formatting for Name is listed on the resource.

For Configuration, it's invocations/{invocation}/configs/{config}.

https://github.com/googleapis/googleapis/blob/4eccaaf48c0ccadc6f98707d3dbe9614d47bb103/google/devtools/resultstore/v2/configuration.proto#L33

File: prow/cmd/crier/main.go
Comment (airbornepony) [reply prow/cmd/crier/main.go:132]: The parallelism here is strictly for separate Invocations for different jobs (and IDs), so we should be ok.

File: prow/resultstore/payload.go
Comment (airbornepony) [reply prow/resultstore/payload.go:45]: In order to keep the client as simple as possible, I separated the batch upload concerns into the writer package (writer.go). writer.New() takes the InvocationId from the Invocation to create the UploadRequest and takes care of all the token exchange. There should be no trace of the batch operation in this file.

Thanks for noting that the InvocationID is not needed in any artifacts; I've removed Payload.invID now.

(Sorry, the earlier comment I wrote here was lost on github somewhere when I pushed my last commit, I think...)

File: prow/resultstore/writer/writer.go
Comment (ashleyrichter) [inline prow/resultstore/writer/writer.go:32]: Our guidance is ~100 for batch size.

**Review by listx — COMMENTED** [review]
/lgtm
/hold

Got some nits but they are not blocking.

Recapping the offline discussion we've had, I think the consensus was that we should try to avoid walking the entire GCS path and instead only upload things 1 level down or have some cap like 50 files max. Not sure how much of that will change this PR, so I'll leave it up to you whether you want to do that here or in a follow-up.

I see some TODO comments here but I assume they are not blocking.

File: prow/cmd/crier/main.go
Comment (airbornepony) [reply prow/cmd/crier/main.go]: Done

File: prow/resultstore/files_test.go
Comment (airbornepony) [reply prow/resultstore/files_test.go]: Done

File: prow/apis/prowjobs/v1/types.go
Comment (airbornepony) [reply prow/apis/prowjobs/v1/types.go:358]: Done

**Review by michelle192837 — COMMENTED** [review]
LGTM (though I'll leave approval to others), others have left excellent comments and I'm not seeing anything beyond those that I'm concerned about.

**Comment by airbornepony** [comment]
Re: GCS tree-walk load/performance concerns

In answer to Linus concerns above and offline discussions, I am seeking solutions that avoid walking the entire artifacts/ tree. It would be best to get some experience with GCS performance by testing; we should get the feature in and start some Alpha testing, and iterate.

In discussion with the ResultStore leads, it turned up that writing the ".../artifacts/" top-level directory may be suitable for the ResultStore GUI to effectively render all the artifacts contained. It doesn't seem to be documented, and since we cannot confirm, we will have to experiment. My most recent commit (9b136a9) adds the ability to write the top-level "artifacts/" directory based on a boolean flag, so we can try both approaches easily.

File: prow/apis/prowjobs/v1/types.go
Comment (cjwagner) [inline prow/apis/prowjobs/v1/types.go:358]: nit: We had some confusion previously about which projectID this should actually be (I think there were potentially 3 projectIDs involved) so it would be helpful to name this more precisely or add a comment that describes precisely which projectID this should be set to.

Comment (cjwagner) [inline prow/apis/prowjobs/v1/types.go:358]: This value is never set because the type is only defined on the ProwJob resource, not in the job config.
Notice how the slack reporter configuration exists in central prow config for defaulting: https://github.com/kubernetes/test-infra/blob/139299a6007c3ef448adad5df389a8d30fdc4517/prow/config/config.go#L152
and also at the job level for job specific overrides: https://github.com/kubernetes/test-infra/blob/139299a6007c3ef448adad5df389a8d30fdc4517/prow/config/jobs.go#L123C5-L123C5

As mentioned before, I don't think we want to follow the same defaulting config pattern as the slack reporter. I think it would be more appropriate to add the fields to the `ProwJobDefault` struct which currently is only used for the `TenantID`: https://github.com/kubernetes/test-infra/blob/139299a6007c3ef448adad5df389a8d30fdc4517/prow/config/config.go#L219-L224C5

**Review by cjwagner — COMMENTED** [review]
I'm still reviewing this, but I noticed that we're missing the configuration side of things for the `projectID` so I wanted to share this feedback now so you can start adding that while I continue reviewing.

File: prow/apis/prowjobs/v1/types.go
Comment (cjwagner) [reply prow/apis/prowjobs/v1/types.go:358]: Discussed offline. The type is actually defined in job config at the job level thanks to the ReporterConfig type being reused and copied as a whole from the config to the resource.
Higher level defaulting will be added in a later PR.

File: prow/resultstore/writer/writer.go
Comment (cjwagner) [reply prow/resultstore/writer/writer.go:92]: The same applies to the retry loop in the `flushUpdates()` function.

File: prow/resultstore/payload.go
Comment (cjwagner) [inline prow/resultstore/payload.go:98]: nit: `prowjob`

File: prow/resultstore/payload.go
Comment (cjwagner) [inline prow/resultstore/payload.go:113]: nit: Should some of these become labels rather than just being in the description? I'm assuming labels are easier to search, filter, or group by.

File: prow/resultstore/payload.go
Comment (cjwagner) [inline prow/resultstore/payload.go:190]: nit: Elsewhere we use `p.Job` without checking if it is nil. We should be consistent about this.

File: prow/resultstore/payload.go
Comment (cjwagner) [inline prow/resultstore/payload.go:182]: nit: Since "default" is used as an identifier repeatedly we should probably make it a constant so the compiler can help us catch any typos.

**Review by cjwagner — COMMENTED** [review]
Looks great overall. I think I noticed some small problems, but nothing that requires a significant change.

One question:
It appears to be possible to specify some metadata fields at many levels (invocation, configuration, configuredtarget, and/or action). Specifically for things like properties, status, and timing. Sometimes the appropriate scope is clear, but for things like the status, I'm curious if we should be setting that once at the highest applicable scope, or at every scope that it applies to.


Future improvment ideas that you probably have already considered:
- Separate actions for each test container. Maybe a separate action for clonerefs (and configure action dependency)?
- Maybe publish some policy for well known artifact names to better support the special post-processing/handling of reserved file IDs other than just build.log.

I did not review the unit tests, I've been staring at this all day and need a break.

File: prow/resultstore/payload.go
Comment (airbornepony) [reply prow/resultstore/payload.go:113]: Many of these Prow labels become resultstore Properties. I believe ResultStore want their Labels to be low-radix since they are indexed. See https://github.com/googleapis/googleapis/blob/2aa3b1d5a90d05e0606d11787de475b0df068d49/google/devtools/resultstore/v2/invocation.proto#L166, so using any kind of ID is bad.

I intend in future work to allow configuration of custom resultstore labels by repo/job. Those are probably the only ones we should create.

File: prow/resultstore/payload.go
Comment (airbornepony) [reply prow/resultstore/payload.go:190]: I'm not finding where we do. Note that the helper functions above for building an Invocation are all guarded by the check in p.invocation().

Note the unit tests exercise these methods with a nil Job (and other nil parts).

In the spirit of nil pointers, I notice now that Spec.Refs and Status.CompletionTime (but not StartTime) are also pointers. Expanded test cases and updated to guard these.

File: prow/resultstore/payload.go
Comment (airbornepony) [reply prow/resultstore/payload.go:98]: Done

File: prow/resultstore/payload.go
Comment (airbornepony) [reply prow/resultstore/payload.go:113]: Many of these Prow labels become resultstore Properties. I believe ResultStore want their Labels to be low-radix since they are indexed. See https://github.com/googleapis/googleapis/blob/2aa3b1d5a90d05e0606d11787de475b0df068d49/google/devtools/resultstore/v2/invocation.proto#L166, so using any kind of ID is bad.

I intend in future work to allow configuration of custom resultstore labels by repo/job. Those are probably the only ones we should create.

**Comment by ashleyrichter** [comment]
> One question: It appears to be possible to specify some metadata fields at many levels (invocation, configuration, configuredtarget, and/or action). Specifically for things like properties, status, and timing. Sometimes the appropriate scope is clear, but for things like the status, I'm curious if we should be setting that once at the highest applicable scope, or at every scope that it applies to.
> 

For Status and Timing, we recommend setting at every level, aggregating up to each level, that will help the UI be more useful.

> Future improvment ideas that you probably have already considered:
> 
> * Separate actions for each test container. Maybe a separate action for clonerefs (and configure action dependency)?
> * Maybe publish some policy for well known artifact names to better support the special post-processing/handling of reserved file IDs other than just build.log.

+1 the separate actions for each test container. I'm not entirely familiar with Prow's data model, but that sounds like how we intend the API to be used.

File: prow/resultstore/files.go
Comment (listx) [reply prow/resultstore/files.go:178]: >  I'm not sure what kind of behavior that would induce in ResultStore.

@ashleyrichter Do you have any insights here on the behavior of ResultStore if we omit the file length and/or the content type?

Comment (ashleyrichter) [reply prow/resultstore/files.go:178]: Length is purely informational for the end user in the UI. If it's not provided, the UI will not show a length rather than try to stat all of the files referenced (as stating arbitrarily many files can be very slow).

`content_type` is used by the UI to determine how the file should be displayed, such as plain text or HTML, etc. If you don't provide the content type, the UI will try to infer it based on file extension or name. If it can't infer content type, it will default to just downloading the file rather than trying to display it in the UI (in other words, it will presume binary data for unknown files).

Comment (listx) [reply prow/resultstore/files.go:178]: Ack.

Hmm,  if we define `content_type` as plain text for everything (assume there is a bug on our end and we do this for all files, text or otherwise), the UI will attempt to read/display binary data? Seems like a possible bug and/or security issue.

In general I'm noticing a pattern of ResultStore trusting clients to always do the right thing, but this does not seem tenable in the long run. Maybe I'm missing something?

File: prow/resultstore/uploader.go
Comment (airbornepony) [reply prow/resultstore/uploader.go]: I've linked what I can find, and asked the team whether there are good public API docs available. I'll update this if/when I find something.

File: prow/resultstore/files.go
Comment (ashleyrichter) [reply prow/resultstore/files.go:178]: I can point you to the internal security review with the confidential details if you reach out to me on Corp, but we treat all file data as potentially hostile and we have protections in place.

In terms of trusting the client, that's complicated and it's a problem we've spent a lot of time on. We encourage users to use a Bazel because it at least makes that problem easier to reason about, but that's not reasonable for all users. Anyways, I can talk more on that too if you reach out to me internally.

Comment (listx) [reply prow/resultstore/files.go:205]: Thanks for clarifying. I don't know what you mean by "We encourage users to use a Bazel..." (perhaps you're saying that if users all used Bazel, then at least the build artifacts could be in some way standardized, which would obviously make ResultStore's life easier).

I don't have any further questions at this time, thanks.

File: prow/apis/prowjobs/v1/types.go
Comment (airbornepony) [reply prow/apis/prowjobs/v1/types.go:365]: Added back documentation.

File: prow/resultstore/client.go
Comment (cjwagner) [reply prow/resultstore/client.go:31]: We'd probably make this a flag for integration testing (with this as the default value). My comment was more specifically about simplifying the process of creating a new client so that callers only need `NewClient` and not `Connect`.

**Review by cjwagner — APPROVED** [review]
Lets give it a go and iterate.

**Pull request merged**
